components/suggest/src/yelp.rs

```diff
@@ -142,69 +153,211 @@ impl<'a> SuggestDao<'a> {
             return Ok(vec![]);
         }
 
-        let query_string = &query.keyword.trim();
-        if !query_string.contains(' ') {
-            let Some((subject, subject_exact_match)) = self.find_subject(query_string)? else {
-                return Ok(vec![]);
-            };
-            let (icon, icon_mimetype, score) = self.fetch_custom_details()?;
-            let builder = SuggestionBuilder {
-                subject: &subject,
-                subject_exact_match,
-                pre_modifier: None,
-                post_modifier: None,
-                location_sign: None,
-                location: None,
-                need_location: false,
-                icon,
-                icon_mimetype,
-                score,
-            };
-            return Ok(vec![builder.into()]);
-        }
+        let query_vec: Vec<_> = query.keyword.split_whitespace().collect();
+        let mut query_words: &[&str] = &query_vec;
 
-        // Find the yelp keyword modifier and remove them from the query.
-        let (query_without_yelp_modifiers, _, _) =
-            self.find_modifiers(query_string, Modifier::Yelp, Modifier::Yelp)?;
+        let pre_yelp_modifier_tuple =
+            self.find_modifier(query_words, Modifier::Yelp, FindFrom::First)?;
+        if let Some((_, n)) = pre_yelp_modifier_tuple {
+            query_words = &query_words[n..];
+        }
 
-        // Find the location sign and the location.
-        let (query_without_location, location_sign, location, need_location) =
-            self.find_location(&query_without_yelp_modifiers)?;
+        let pre_modifier_tuple = self.find_modifier(query_words, Modifier::Pre, FindFrom::First)?;
+        if let Some((_, n)) = pre_modifier_tuple {
+            query_words = &query_words[n..];
+        }
 
-        if let (Some(_), false) = (&location, need_location) {
-            // The location sign does not need the specific location, but user is setting something.
+        let subject_tuple = self.find_subject(query_words)?;
+        if let Some((_, _, n)) = subject_tuple {
+            query_words = &query_words[n..];
+        } else {
             return Ok(vec![]);
         }
```

Comment:
You can use [let-else](https://doc.rust-lang.org/rust-by-example/flow_control/let_else.html) to avoid having to unwrap `subject_tuple` below (on line 201):

```suggestion
        if let Some(subject_tuple) = self.find_subject(query_words)? else {
            return Ok(vec![]);
        };
        // Now you can use subject_tuple without unwrapping it
        query_words = &query_words[subject_tuple.2..]
```

Reply:
Oh, nice, it is more readable. Thank you so much!